### ISSUE TITLE
Yolo12 batch size

### ISSUE THREAD

**Issue by RobbStarkAustria** [ultralytics/ultralytics#19488]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Hi,

I have a question about the batch size when using yolo12. I want to compare my results of a yolo11l-model trained by my own dataset. Furthermore, I use this setting:

```
{'exist_ok': True,
'cache': 'disk',
'workers': 12,
'batch': 20,
'patience': 10,
'epochs': 65,
'device': 0,
'imgsz': 1408,
'rect': True,
'optimizer': 'AdamW',
'max_det': 1000,
'lr0': 0.0001,
'cos_lr': True,
'augment': False,
'plots': True,
'cls': 0.5,
'dfl': 1.5,
'nbs': 64,
'hsv_h': 0.0,
'hsv_s': 0.0,
'hsv_v': 0.0,
'degrees': 0,
'translate': 0,
'scale': 0,
'shear': 0,
'perspective': 0,
'flipud': 0,
'fliplr': 0,
'mosaic': 1.0,
'mixup': 0,
'copy_paste': 0.0,
'auto_augment': None,
'erasing': 0.0,
'crop_fraction': 0.0,
'name': 'sd_st_sc_g_12l_25_03_01_no_shadow'}
```

All the train-images have the size of 1408x192, so I use the image-size 1408 and the option rect=True. As you see, I can do training with a batch-size of 20 with no memory issues. I'm using this system:
```
Ultralytics 8.3.81 🚀 Python-3.12.8 torch-2.5.1 CUDA:0 (NVIDIA GeForce RTX 4070 SUPER, 12281MiB)
Setup complete ✅ (24 CPUs, 50.0 GB RAM, 350.9/1006.9 GB disk)
```

If I change the starting model for transfer-training from yolo11l to yolo12l all GPU-memory is consumed, and even with a batch-size of 1 to get no reasonable GPU-utilization. I noticed that the batch-size calculated by autobatch if using the following test-system is also decreased from 14 (yolo11l) to 7 (yolo12l).

```
import ultralytics
from ultralytics import YOLO, settings
ultralytics.checks()
print(settings)
model = YOLO("yolo11l.pt")  # load a pretrained model (recommended for training)
# Use the model
model.train(data='coco8.yaml', epochs=3, auto_augment=None, batch=0.8)  # train the model)
```

So I think that lowering the batch size should be necessary. But to set the batch-size to 1 increases the training time too much.

So, what happened? Maybe the parameter rect is not right interpreted?

Thank you for your answers!

Kind regards

Olaf

**Comment by UltralyticsAssistant** [comment]
👋 Hello @RobbStarkAustria, thank you for your interest in Ultralytics 🚀! 

We recommend starting with a visit to the [Docs](https://docs.ultralytics.com/) where you can find guidance on configurations and training, including details on [Python](https://docs.ultralytics.com/usage/python/) and [CLI](https://docs.ultralytics.com/usage/cli/) usage examples. Many common questions may already be addressed there.

If this is a 🐛 Bug Report or an unexpected behavior, it would be very helpful if you could provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that highlights the issue clearly. This helps us debug and address the problem more effectively.

If you're facing challenges related to custom training ❓, please provide as much detail as possible, including:
- Dataset characteristics (e.g., image size, format, and sample examples)
- Training logs
- Verification that you are following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/)

## Upgrade

To ensure you're using the most optimized version, please upgrade to the latest `ultralytics` package along with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml). This can be done within a [**Python>=3.8**](https://www.python.org/) environment and [**PyTorch>=1.8**](https://pytorch.org/get-started/locally/):

```bash
pip install -U ultralytics
```

## Debugging and Resources

You can also use our runtime-tested options for deploying YOLO in the following environments that have pre-installed dependencies ([CUDA](https://developer.nvidia.com/cuda-zone), [CUDNN](https://developer.nvidia.com/cudnn), [Python](https://www.python.org/), [PyTorch](https://pytorch.org/)):
- **Jupyter and Notebooks**: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Cloud Platforms**: Refer to the [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/) or [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Containerization**: Use our [Docker Image](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

For real-time interaction, make sure to check out:
- [Discord](https://discord.com/invite/ultralytics) 🎧
- [Discourse forums](https://community.ultralytics.com/) for in-depth discussions
- [Subreddit](https://reddit.com/r/Ultralytics) to connect with our community.

## Status and Support

If you're using YOLO12 and noticing differences in autobatch calculations or resource utilization, it may be related to the architecture differences between YOLO11 and YOLO12 models. For further investigation, we suggest:
- Double-check the `nbs` (Nominal Batch Size) parameter to match the newer model structure
- Experiment with floating-point precision (e.g., using `torch.set_float32_matmul_precision('high')` for optimization)
- Review GPU memory allocation on your system for potential conflicts or overhead

Finally, keep an eye on our automated CI tests, which ensure the integrity of YOLO across platforms:
<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a> 

An Ultralytics engineer will review this issue soon for further insight. Thank you for your patience and detailed report. Your engagement helps improve YOLO for everyone 🚀!

**Comment by Y-T-G** [comment]
YOLO12 uses attention. It requires more memory.

**Comment by RobbStarkAustria** [comment]
I'm wondering if this means I can only use yolo12 for my own data set on my system, with some restrictions.

Or are there any other options, as the memory is also filled with batch-size 1?

**Comment by Y-T-G** [comment]
You can use a smaller model. Like m. Or smaller imgsz.

**Comment by glenn-jocher** [comment]
@RobbStarkAustria I'd recommend using YOLO11. 12 has some unresolved memory and stability issues, and it's mainly only useful for R&D if you have a large budget and lots of time to spend debugging.

**Comment by glenn-jocher** [comment]
For accuracy-focused applications, we recommend using **YOLO11x** (our largest model) with your dataset. While YOLO12 offers marginal accuracy gains (~+0.4% mAP vs YOLO11l per the YOLO12 documentation), its current memory demands and training instability make it less practical. YOLO11x achieves better accuracy than YOLO12l with more efficient resource usage. You can find implementation details in the YOLO11 documentation. 

For those seeking maximum accuracy, ensure you're using: 
1. Latest `ultralytics>=8.3.81` 
2. FP32 precision (avoid mixed precision)
3. Full image resolution (1408px in your case)

**Comment by RobbStarkAustria** [comment]
@glenn-jocher 
Hi Glenn,

thank you for your advice. I want to inform you about my results. Besides your options, I used a custom dataloader which allows shuffling on my rectangular dataset. This increases the accuracy of the detection from non-shuffling. The table shows the results based on my test-dataset with 22.267 ROIs. All trainings are done with an image size of 1408x192px.

| Settings                               | accuracy (weighted) | F1-score (weighted) | accuracy | F1-Score |
|----------------------------------------|---------------------|---------------------|----------|----------|
| yolo11l, mixed precision, no shuffling |       0.94764       |       0.972128      |  0.8599  | 0.904477 |
| yolo11l, mixed precision, shuffling    |       0.987744      |       0.99198       | 0.937162 | 0.949076 |
| yolo11x, FP32, shuffling               |       0.994884      |       0.996191      | 0.959665 | 0.962806 |

You can see a dramatic increase with shuffling enabled. And your mentions take an extra boost to the accuracy as well.

Why is shuffling disabled by default when rect is true?

**Comment by glenn-jocher** [comment]
Thanks for sharing your interesting findings! Shuffling is disabled by default with `rect=True` to maintain sorted aspect ratios for optimal batch packing efficiency. While this sorting improves speed, your results demonstrate that dataset-specific shuffling can sometimes enhance model performance at the expense of efficiency - a valuable tradeoff insight for certain use cases. 

For those interested in rectangular training fundamentals, see the [Ultralytics rectangular training documentation](https://docs.ultralytics.com/guides/model-training-tips/#rectangular-training). We'll consider making shuffling configurable in future updates.

**Comment by xuan-xuan6** [comment]
> [@glenn-jocher](https://github.com/glenn-jocher) Hi Glenn,
> 
> thank you for your advice. I want to inform you about my results. Besides your options, I used a custom dataloader which allows shuffling on my rectangular dataset. This increases the accuracy of the detection from non-shuffling. The table shows the results based on my test-dataset with 22.267 ROIs. All trainings are done with an image size of 1408x192px.
> 
> Settings	accuracy (weighted)	F1-score (weighted)	accuracy	F1-Score
> yolo11l, mixed precision, no shuffling	0.94764	0.972128	0.8599	0.904477
> yolo11l, mixed precision, shuffling	0.987744	0.99198	0.937162	0.949076
> yolo11x, FP32, shuffling	0.994884	0.996191	0.959665	0.962806
> You can see a dramatic increase with shuffling enabled. And your mentions take an extra boost to the accuracy as well.
> 
> Why is shuffling disabled by default when rect is true?
> 
> If you are interested in my use-case, check my github-repo: [https://github.com/RobbStarkAustria/mensural_to_mei](url)

hello，I‘d like to consult you how to make Shuffling?any code?
